runners/spark/src/main/java/org/apache/beam/runners/spark/translation/ValueAndCoderKryoSerializable.java

```diff
@@ -0,0 +1,203 @@
+ */
+package org.apache.beam.runners.spark.translation;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.KryoSerializable;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.google.common.base.Throwables;
+import com.google.common.io.ByteStreams;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.OutputStream;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.util.VarInt;
+import org.apache.beam.sdk.util.common.ElementByteSizeObserver;
+
+/**
+ * A holder object that lets you serialize an element with a Coder with minimal wasted space.
+ * Supports both Kryo and Java serialization.
+ *
+ * <p>There are two different representations: a deserialized representation and a serialized
+ * representation.
+ *
+ * <p>The deserialized representation stores a Coder and the value. To serialize the value, we write
+ * a length-prefixed encoding of value, but do NOT write the Coder used.
+ *
+ * <p>The serialized representation just reads a byte array - the value is not deserialized fully.
+ * In order to get at the deserialized value, the caller must pass the Coder used to create this
+ * instance via getOrDecode(Coder). This reverts the representation back to the deserialized
+ * representation.
+ *
+ * @param <T> element type
+ */
+public class ValueAndCoderKryoSerializable<T> implements KryoSerializable, Externalizable {
+  private T value;
+  // Re-use a field to save space in-memory. This is either a byte[] or a Coder, depending on
+  // which representation we are in.
+  private Object coderOrBytes;
+
+  ValueAndCoderKryoSerializable(T value, Coder<T> currentCoder) {
+    this.value = value;
+    this.coderOrBytes = currentCoder;
+  }
+
+  @SuppressWarnings("unused") // for serialization
+  public ValueAndCoderKryoSerializable() {}
+
+  public T getOrDecode(Coder<T> coder) throws IOException {
+    if (!(coderOrBytes instanceof Coder)) {
+      value =
+          coder.decode(new ByteArrayInputStream((byte[]) this.coderOrBytes), Coder.Context.OUTER);
+      this.coderOrBytes = coder;
+    }
+
+    return value;
+  }
+
+  private static class ByteSizeObserver extends ElementByteSizeObserver {
+    private long observedSize = 0;
+
+    @Override
+    protected void reportElementSize(long elementByteSize) {
+      observedSize += elementByteSize;
+    }
+  }
+
+  private void writeCommon(OutputStream out) throws IOException {
+    if (!(coderOrBytes instanceof Coder)) {
+      byte[] bytes = (byte[]) coderOrBytes;
+      VarInt.encode(bytes.length, out);
+      out.write(bytes);
+    } else {
+      @SuppressWarnings("unchecked")
+      Coder<T> coder = (Coder<T>) coderOrBytes;
+      int bufferSize = 1024;
+
+      if (coder.isRegisterByteSizeObserverCheap(value)) {
```

Comment:
Usually. `CoderHelpers` are utilized in spark runner for `Object` to bytes (and vice versa) conversions in spark runner. This is a first usage of  `ByteSizeObserver` in spark runner. Would it be worth to move serialization using `ByteSizeObserver` to  `CoderHelpers`? @iemejia

Reply:
I'm not 100% sure if this matters, but aside from the observer, this code doesn't actually ever produce a `byte[]` of the serialized output. Specifically, `ByteArrayOutputStream.writeTo` is used to avoid that. If the buffer inside `ByteArrayOutputStream` is bigger than the content, this allows us to avoid an array copy.

To avoid the copy properly when returning bytes, Java has a `ByteBuffer`. I'd have to change `CoderHelpers` to use `ByteBuffer`s instead - and I'm not sure if all consumers really benefit from the unwieldy API that gives them. My suggestion in this case would be - if you want performance in Spark serialization, you should use `ValueAndCoderLazySerializable` directly. Otherwise, use `CoderHelpers` for explicit serialization/deserialization as needed.